crates/assists/src/handlers/extract_function.rs

```diff
@@ -1053,17 +1147,42 @@ fn make_ret_ty(ctx: &AssistContext, module: hir::Module, fun: &Function) -> Opti
             make::ty_generic(make::name_ref("Option"), iter::once(handler_ty))
         }
         FlowHandler::MatchOption { .. } => {
-            make::ty_generic(make::name_ref("Option"), iter::once(ty.make_ty(ctx, module)))
+            make::ty_generic(make::name_ref("Option"), iter::once(fun_ty.make_ty(ctx, module)))
         }
         FlowHandler::MatchResult { err } => {
             let handler_ty =
                 err.expr_ty(ctx).map(|ty| make_ty(&ty, ctx, module)).unwrap_or_else(make::ty_unit);
-            make::ty_generic(make::name_ref("Result"), vec![ty.make_ty(ctx, module), handler_ty])
+            make::ty_generic(
+                make::name_ref("Result"),
+                vec![fun_ty.make_ty(ctx, module), handler_ty],
+            )
         }
     };
     Some(make::ret_type(ret_ty))
 }
 
+/// Extract `E` type from `Result<T, E>`
+fn result_err_ty(
+    parent_ret_ty: &hir::Type,
+    ctx: &AssistContext,
+    module: hir::Module,
+) -> Option<ast::Type> {
+    // FIXME: use hir to extract argument information
+    //        currently we use `format -> take part -> parse`
+    let path_ty = match make_ty(&parent_ret_ty, ctx, module) {
+        ast::Type::PathType(path_ty) => path_ty,
+        _ => return None,
+    };
+    let arg_list = path_ty.path()?.segment()?.generic_arg_list()?;
+    let err_arg = arg_list.generic_args().nth(1)?;
+    let type_arg = match err_arg {
+        ast::GenericArg::TypeArg(type_arg) => type_arg,
+        _ => return None,
+    };
+
+    type_arg.ty()
```

Comment:
I haven't found a way to type arguments from `hir::Type` instance of `Result<T, E>`. Have I missed something?

Reply:
I think you should be good to add a function to `hir::Type` that walks only substitutions of a type and that only once which should allow you to query `T` and `E` in this case I think?
See https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/hir/src/code_model.rs#L1896

Reply:
If it's just about getting the type parameters, there's no reason to make it use a callback, it can just return an iterator. I'm not sure what the goal here is though?

Reply:
When extracting function with try `expr?`, it looks up return type of outer function, e.g. `Result<(), Error>`, and computes return type of the generated function if we forget about try `expr?`, e.g. `i32`. Then the actual return type of generated function is `Result<i32, Error>`. The goal of this function is to get the `Error` type from the `Result<(), Error>`

If I understood you correctly, it is okay to create `fn substitutions(&self) -> &Substs` in `hir::Type`?

Reply:
`fn type_parameters(&self) -> impl Iterator<Item = hir::Type>` is what we want there. I don't think `Substs` should be exposed from the `hir` crate.